DEV-93860 added config to support deviceless logs and use metadata.

siddharthck opened the pull request:
Added config to exclude metadata fields. updated README.md
bumping up patch to 1.0.2

ahsan13jan commented:
We should have some unit tests.

ahsan13jan reviewed (changes requested):
Need some unit tests.

siddharthck commented:
> Need some unit tests.

I have added unit tests for for device_less_logs, include_metadata and metadata_to_exclude combinations

ahsan13jan reviewed (approved):
LGTM

ahsan13jan reviewed (changes requested):
Deviceless logs, needs service and namespace.

siddharthck commented:
> Deviceless logs, needs service and namespace.

@ahsan13jan , **_service_** and **_namespace_** would be different for different type of logs. If we could define service and namespace in **_record_transformer_**, It will always get populated when **_device_less_logs_** is set to **_true_** unless defined as part of _**metadata_to_exclude []**_.  It can be solved with configuration.

ahsan13jan commented:
@siddharthck our plugin does not know about record_transformer,

siddharthck commented:
> @siddharthck our plugin does not know about record_transformer,

@ahsan13jan I have added an hard condition that record must have 'service' and 'namespace' when device_less_logs is 'true'. Also updated the doc.

ahsan13jan commented on lib/fluent/plugin/out_lm.rb:
```diff
@@ -102,24 +107,34 @@ def process_record(tag, time, record)
       end
 
-      if @include_metadata
+      if @include_metadata || @device_less_logs
```
Why @device_less_logs is there in this condition?

ahsan13jan commented on test/plugin/test_out_lm.rb:
```diff
@@ -147,5 +147,97 @@ def create_valid_subject
 
       assert_equal expected, result
     end
+
```
Nit: We should have tests for deviceless logs in a separate file, so that each feature get's its own file for better readability.

ahsan13jan commented on test/plugin/test_out_lm.rb:
```diff
@@ -147,5 +147,97 @@ def create_valid_subject
+      ]).instance
+      tag = "lm.test"
+      time = Time.parse("2020-08-23T00:53:15+00:00").to_i
+      record = {"message" => "Hello from test",  "timestamp" => "2020-10-30T00:29:08.629701504Z" , "meta1" => "testMeta1" , "meta2" => "testMeta2", "meta3" => "testMeta3", "meta4" => "testMeta4", "meta5" => {"key1" => "value1", "key2" => {"key2_1" => "value2_1"}}, "someProp" => "resourcePropVal" }
```
Nit: Have it in multiline as expected msg is for better readibility.

ahsan13jan commented on test/plugin/test_out_lm.rb:
```diff
@@ -147,5 +147,97 @@ def create_valid_subject
+
+      result = plugin.process_record(tag, time, record)
+        
+      expected = nil
```
When we return nil, I don't think we have a nil check on line: 80

ahsan13jan commented on test/plugin/test_out_lm.rb:
```diff
@@ -147,5 +147,97 @@ def create_valid_subject
+      plugin = create_driver(%[
+        resource_mapping {"a.b": "lm_property"} 
+        device_less_logs true
+        include_metadata false
+        metadata_to_exclude ["meta1", "meta3"]
```
if metadata is disabled, then why are still sending metadata?
Then metadata_to_exclude should be irrelevant.

ahsan13jan reviewed (approved):
LGTM
